### DISCUSSION TITLE
CollisionQueries don't detect deep intersections

### DISCUSSION THREAD

**Discussion by maxwave11** [bepu/bepuphysics2#328] [Q&A] [answered]
Hi!

I'm considering to use this library for clash detection (no physics, no dynamics), just collsion checks between arbitrary meshes.
In CollisionQueryDemo I've noticed this comment:

```C#
// Add in a static object to test against. Note that the mesh triangles are one sided, 
// so some of the queries whose centers are beneath the mesh do not generate any contacts.
var mesh = DemoMeshHelper.CreateDeformedPlane(20, 20, (x, y) => { return new Vector3(x * 5 - 50, 3 * MathF.Sin(x) * MathF.Sin(y), y * 5 - 50); }, Vector3.One, BufferPool);
Simulation.Statics.Add(new StaticDescription(new Vector3(0, 2, 0), Simulation.Shapes.Add(mesh)));
```

![image](https://github.com/bepu/bepuphysics2/assets/1698678/ba4fdb89-8a53-4e6e-b581-357e37774314)

So, is it possible somehow to force batcher include all physically intersected meshes? Also I need to understand how deep one mesh is inside anoher (negative distance).

**Comment by RossNordby** [comment, marked as answer]
The mesh could be made double sided by duplicating the triangles and flipping the winding on one of the duplicates, but that will have other consequences (like it being easier for stuff to get stuck behind walls).

Mesh-mesh collision will not give you a very useful estimate of penetration distance; meshes are composed of infinitely thin one-sided triangles, and the contacts they produce reflect that. That's part of the reason why I frequently and strongly recommend against simulation configurations that would permit mesh-mesh collisions.

Robust solid meshes (or an equivalently expressive solid concave representation) are on the longer-term to-do list. V1 had them, but I never liked the implementation. I wouldn't wait on me to get it done, though, I'm already pretty swamped.

**Reply by Kermalis** [reply]
I was going crazy trying to deal with this all day. Essentially I'm making "checkpoints" that a box shape must collide with. I basically set up the same ContactDemos in my project but also added the ability for a static to not collide with things, so it works as a trigger. This worked perfectly if the checkpoint was using a box shape too, but today I wanted to change the checkpoint to a mesh shape.
So I encountered this behavior where the box would get halfway through the triangles of the checkpoint mesh and then stop reporting collisions. Your suggestion of duplicating + flipping the verts solved all the issues.
Of course, being fully inside the checkpoint mesh reports no collisions (unlike if the checkpoint is a box shape), but I expected that since every physics engine has that behavior and it's an impossible thing to expect. For me this is the perfect solution, so thank you as always. I am assuming that in my case, there isn't much downside to this approach 😃 

~~Edit: Actually, I'm finding that this method doesn't detect collisions on the first tick that they should be in contact, depending on the angle. I may end up needing to go with a standard box shape, which isn't a huge deal. But this method works if you do want to just detect the collisions, even if it's not right away~~
Edit2: It seems I'm getting the contact events 1 tick late with the boxes also, so it must be my implementation instead of bepuphysics. Perhaps both methods will work perfectly, so ignore my first edit
